clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1053,6 +1121,54 @@ static target getAccessTarget(QualType FieldTy,
       AccTy->getTemplateArgs()[3].getAsIntegral().getExtValue());
 }
 
+static bool IsFreeFunction(Sema &SemaRef, const FunctionDecl *FD) {
+  if (FD->hasAttr<SYCLAddIRAttributesFunctionAttr>()) {
+    const auto &Attrs = FD->getAttrs();
+    for (unsigned i = 0, e = Attrs.size(); i < e; i++) {
+      const SYCLAddIRAttributesFunctionAttr *AnAttr =
+          dyn_cast<SYCLAddIRAttributesFunctionAttr>(Attrs[i]);
+      if (AnAttr) {
+        SmallVector<std::pair<std::string, std::string>, 4> NameValuePairs =
+            AnAttr->getAttributeNameValuePairs(SemaRef.Context);
+        for (const auto &NameValuePair : NameValuePairs) {
+          if (NameValuePair.first == "sycl-range-kernel" ||
+              NameValuePair.first == "sycl-nd-range-kernel" ||
+              NameValuePair.first == "sycl-single-task-kernel")
+            return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+static std::string constructFFKernelName(ASTContext &Ctx,
+                                         const FunctionDecl *FD) {
+  IdentifierInfo *Id = FD->getIdentifier();
+  std::string NewIdent = (Twine("__free_function_") + Id->getName()).str();
+  return NewIdent;
+}
+
+// Gets a name for the free function kernel function. The suffix allows a normal
+// device function to coexist with the kernel function.
+static std::pair<std::string, std::string>
+constructFreeFunctionKernelName(Sema &S, const FunctionDecl *KernelCallerFunc,
+                                MangleContext &MC) {
+  SmallString<256> Result;
+  llvm::raw_svector_ostream Out(Result);
+  std::string MangledName;
+  std::string StableName;
+
+  if (KernelCallerFunc->getTemplateSpecializationArgs()) {
```

Comment:
Can you explain what it happening here? If the free function is a function template you're mangling the name and appending a prefix but if it isn't you're just appending a prefix?

Reply:
Also please add tests for naming.

Reply:
The spec states that a free function should be callable from device code as any other device function and should also be a kernel function callable from the host. So each user-written free function is placed in device code as two different function (mainly differing in the parameter list). We have to give the two versions different names (whether they were templated or not). I've added the string "__free_function_" to the kernel function to distinguish it from the normal device function.

Reply:
Added a templated free function test-case to the AST lit test "free_function_kernel_params.cpp" and the names of plain and templated free functions are now checked.